src/modules/hooks/base_hooks.js

```diff
@@ -0,0 +1,55 @@
+module.exports = class BaseHooks {
+  constructor() {
+    this._events = {};
+  }
+
+  _on(event, callback) {
+    if (Array.isArray(callback)) {
+      this._manyOn(event, callback);
+      return;
+    }
+
+    if (!this._events[event]) {
+      this._events[event] = new Set();
+    }
+    this._events[event].add(callback);
+  }
+
+  _manyOn(event, callbacks) {
+    if (!this._events[event]) {
+      this._events[event] = new Set();
+    }
+    callbacks.forEach((callback) => this._events[event].add(callback));
+  }
+
+  _off(event, callback) {
+    if (Array.isArray(callback)) {
+      this._manyOff(event, callback);
+      return;
+    }
+
+    if (!this._events[event]) {
+      return;
+    }
+    this._events[event].delete(callback);
+  }
+
+  _manyOff(event, callbacks) {
+    if (!this._events[event]) {
+      return;
+    }
+    callbacks.forEach((callback) => this._events[event].delete(callback));
+  }
+
+  async trigger(source, eventName, payload) {
+    if (!this._events[eventName]) {
+      return;
+    }
+
+    // This rule intends to restrict it for arrays, but this is a Set which doesn't have a '.map' function to use instead.
```

Comment:
Suggest to have better failure handling here:
* Since we don't want a hook to fail a process (openning connection, sending message) - so it means all places that uses `.trigger` will need to try..catch
* Do we want that one failure of hook will cause other hooks to not sure?

Maybe use `Promise.allSettled` and log if there any errors?

Reply:
I wasn't sure how to handle this.
I thought it is reasonable to delegate the error handling duty to the user who registers the callbacks, so they can decide whether to fail an operation or not.
I also think there is value in calling the callbacks one by one instead of a `Promise.all`-like action, in case the user wants to register callbacks to be called in a specific order.
Lastly, I wasn't sure if the callbacks should be asynchroneous at all.

So I implemented it in a way that I thought gives the user the most versatility, but also the most responsiblilty.
I'll add in the docs comments guidelines to the callback registration.

Reply:
Actually since I save the callbacks in a `Set` they won't always execute in the same order, so that point doesn't mind.

Reply:
I disagree here, I see those hooks as observers and not as interceptors, so I don't except them to break existing flow.
As I don't expect otel instrumentation to break publish/consume because of it's errors.

If we don't handle it here, it means that all implemented hooks needs to have `try..catch` around them and make sure they are safe.

Reply:
Like discussed, wrapped the hook trigger with try/catch, and allowed the hooks to return `false` to cancel operations when applicable.